### PR TITLE
[Merged by Bors] - Env var to ignore ports used by debugger for intelliJ

### PR DESCRIPTION
**Pull request by infiniteregrets**
refer #644

### CODE REVIEW

File: mirrord-layer/src/lib.rs [hunk 1]
```diff
@@ -117,6 +117,23 @@ fn is_nix_or_devbox() -> bool {
     }
 }
 
+/// Prevent mirrord from connecting to ports used by the intelliJ debugger
+pub(crate) fn intellij_debug_patch(port: Port) -> bool {
+    if let Ok(ports) = std::env::var("INTELLIJ_DEBUG_PATCH") {
+        // port range can be specified as "1234-1236" or just "1234"
+        let ports: Vec<u16> = ports.split('-').map(|p| p.parse().unwrap()).collect();
```
Comment (meowjesty) [inline mirrord-layer/src/lib.rs:124]: Change `unwrap` here to `expect`, so you can give a nicer user error, like "Tried to parse a value that is not a number" or something along those lines.

File: intellij-ext/src/main/kotlin/com/metalbear/mirrord/MirrordDefaultConfig.kt [hunk 2]
```diff
@@ -19,6 +19,7 @@ data class MirrordDefaultConfig(
     val rustLog: LogLevel = LogLevel.INFO,
     val overrideEnvVarsExclude: String = "",
     val overrideEnvVarsInclude: String = "*",
+    val ignorePorts: String = "",
```
Comment (meowjesty) [inline intellij-ext/src/main/kotlin/com/metalbear/mirrord/MirrordDefaultConfig.kt:22]: Considering how lacking the documentation is around this, will the user even know which ports to ignore? Maybe we should do a default range for intellij?

**Comment by jamowei** [comment]
Nice work! I would recommend using a more general name for the new env var instead of `INTELLIJ_IGNORE_PORTS`. Because in my opinion, this is not a problem with IntelliJ itself and also for other use-cases a valid option to choose. My suggestion is something like `MIRRORD_SKIP_PORTS` like `MIRRORD_SKIP_PROCESSES`.

File: intellij-ext/src/main/kotlin/com/metalbear/mirrord/MirrordDefaultConfig.kt (hunk 2 above)
Comment (infiniteregrets) [reply intellij-ext/src/main/kotlin/com/metalbear/mirrord/MirrordDefaultConfig.kt:22]: sure, but it is kind of random. we can keep it from 45K-65K wdyt?

**Comment by meowjesty** [comment]
@jamowei 

 > MIRRORD_SKIP_PORTS

Yeah we plan on adding something like this in the (near) future, just gotta figure out a good API for this first (the user should also be able to specify ip:port combination, port range + some individual specific ports, maybe even have some default skips based on the debugger - like the intellij plugin checking the port automatically).

That's why we're (initially) committing for a small intellij fix.

**Review by meowjesty — APPROVED** [review]
👍

File: CHANGELOG.md [hunk 3]
```diff
@@ -19,6 +19,10 @@ Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how
 - mirrord-config: Changed the way [`fs`](mirrord-config/src/fs.rs) works: now it supports 2 modes `Simple` and `Advanced`, where `Simple` is similar to the old behavior (enables read-only, read-write, or disable file ops), and `Advanced` allows the user to specify include and exclude (regexes) filters for [`FileFilter`](mirrord-layer/src/file/filter.rs).
 - Lint `README` and update it for `--target` flag.
 
+### Fixed
+
+- Env var to ignore ports used by a debugger for intelliJ, refer [#644](https://github.com/metalbear-co/mirrord/issues/644)
+
```
Comment (meowjesty) [inline CHANGELOG.md:25]: Add a notice about the regression you fixed as well with `is_ignored_ip`.

File: mirrord-layer/src/socket.rs [hunk 4]
```diff
@@ -129,6 +129,12 @@ const fn is_ignored_port(port: Port) -> bool {
     port == 0 || (port > 50000 && port < 60000)
 }
 
+/// Ignore localhost Ipv4 and Ipv6 addresses, used along with `is_ignored_port` for debuggers
+#[inline]
+fn is_ignored_ip(ip: IpAddr) -> bool {
```
Comment (aviramha) [inline mirrord-layer/src/socket.rs:134]: why not mix it with the `port_debug_patch`? those ips aren't generally ignored and might confuse reader

**Review by aviramha — CHANGES_REQUESTED** [review]
Almost there.
Shouldn't be a change to the IntelliJ extension to use the new flag?

File: mirrord-layer/src/lib.rs [hunk 5]
```diff
@@ -117,6 +118,30 @@ fn is_nix_or_devbox() -> bool {
     }
 }
 
+/// Prevent mirrord from connecting to ports used by the intelliJ debugger
+pub(crate) fn port_debug_patch(addr: SocketAddr) -> bool {
+    if let Ok(ports) = std::env::var("DEBUGGER_IGNORE_PORTS_PATCH") {
+        let (ip, port) = (addr.ip(), addr.port());
+        let ignored_ip =
```
Comment (aviramha) [inline mirrord-layer/src/lib.rs:125]: nit pick - can just return false here in case of not ignored ip.

File: mirrord-layer/src/lib.rs [hunk 6]
```diff
@@ -117,6 +118,30 @@ fn is_nix_or_devbox() -> bool {
     }
 }
 
+/// Prevent mirrord from connecting to ports used by the intelliJ debugger
+pub(crate) fn port_debug_patch(addr: SocketAddr) -> bool {
+    if let Ok(ports) = std::env::var("DEBUGGER_IGNORE_PORTS_PATCH") {
```
Comment (eyalb181) [inline mirrord-layer/src/lib.rs:123]: Doesn't Rust have some method that would let you unwrap_or_return(false)?

File: mirrord-layer/src/socket.rs [hunk 7]
```diff
@@ -125,8 +125,10 @@ pub struct UserSocket {
 }
 
 #[inline]
-const fn is_ignored_port(port: Port) -> bool {
-    port == 0 || (port > 50000 && port < 60000)
+fn is_ignored_port(addr: SocketAddr) -> bool {
+    let (ip, port) = (addr.ip(), addr.port());
+    let ignored_ip = ip == IpAddr::V4(Ipv4Addr::LOCALHOST) || ip == IpAddr::V6(Ipv6Addr::LOCALHOST);
+    port == 0 || ignored_ip && (port > 50000 && port < 60000)
```
Comment (eyalb181) [inline mirrord-layer/src/socket.rs:131]: Aren't these ports the VSCode debugger ports, and thus should be passed from the VS Code extension?

**Review by aviramha — APPROVED** [review]
bors r+

**Pull request closed without merging**